rubble/src/att/mod.rs

```diff
@@ -135,16 +135,370 @@ enum AttMsg<'a> {
+        attribute_type: AttUuid,
+    },
+    ReadByTypeRsp {
+        /// The length of each attribute handle-value pair in the list
+        length: u8,
+        data_list: HexSlice<&'a [u8]>,
+    },
+    ReadReq {
+        handle: AttHandle,
+    },
+    ReadRsp {
+        value: HexSlice<&'a [u8]>,
+    },
+    ReadBlobReq {
+        handle: AttHandle,
+        offset: u16,
+    },
+    ReadBlobRsp {
+        value: HexSlice<&'a [u8]>,
+    },
+    ReadMultipleReq {
+        /// Minimum length of two handles
+        handles: HexSlice<&'a [u8]>,
+    },
+    ReadMultipleRsp {
+        values: HexSlice<&'a [u8]>,
+    },
     ReadByGroupReq {
         handle_range: RawHandleRange,
         group_type: AttUuid,
     },
+    ReadByGroupRsp {
+        length: u8,
+        data_list: HexSlice<&'a [u8]>,
+    },
+    WriteReq {
+        handle: AttHandle,
+        value: HexSlice<&'a [u8]>,
+    },
+    WriteRsp,
+    WriteCommand {
+        handle: AttHandle,
+        value: HexSlice<&'a [u8]>,
+    },
+    SignedWriteCommand {
+        handle: AttHandle,
+        value: HexSlice<&'a [u8]>,
+        signature: HexSlice<&'a [u8]>, //HexSlice<[u8; 12]>,
+    },
+    PrepareWriteReq {
+        handle: AttHandle,
+        offset: u16,
+        value: HexSlice<&'a [u8]>,
+    },
+    PrepareWriteRsp {
+        handle: AttHandle,
+        offset: u16,
+        value: HexSlice<&'a [u8]>,
+    },
+    ExecuteWriteReq {
+        /// 0x00 – Cancel all prepared writes
+        /// 0x01 – Immediately write all pend- ing prepared values
+        flags: u8,
+    },
+    ExecuteWriteRsp,
+    HandleValueNotification {
+        handle: AttHandle,
+        value: HexSlice<&'a [u8]>,
+    },
+    HandleValueIndication {
+        handle: AttHandle,
+        value: HexSlice<&'a [u8]>,
+    },
+    HandleValueConfirmation,
     Unknown {
         opcode: Opcode,
         params: HexSlice<&'a [u8]>,
     },
 }
 
+impl<'a> AttMsg<'a> {
+    fn from_reader(bytes: &mut ByteReader<'a>, opcode: Opcode) -> Result<Self, Error> {
+        let auth = opcode.is_authenticated();
+
+        let msg = match opcode {
+            Opcode::ErrorRsp => AttMsg::ErrorRsp {
+                opcode: Opcode::from(bytes.read_u8()?),
+                handle: AttHandle::from_bytes(bytes)?,
+                error_code: ErrorCode::from(bytes.read_u8()?),
+            },
+            Opcode::ExchangeMtuReq => AttMsg::ExchangeMtuReq {
+                mtu: bytes.read_u16_le()?,
+            },
+            Opcode::ExchangeMtuRsp => AttMsg::ExchangeMtuRsp {
+                mtu: bytes.read_u16_le()?,
+            },
+            Opcode::FindInformationReq => AttMsg::FindInformationReq {
+                handle_range: RawHandleRange::from_bytes(bytes)?,
+            },
+            Opcode::FindInformationRsp => AttMsg::FindInformationRsp {
+                format: bytes.read_u8()?,
+                data: HexSlice(bytes.read_slice(bytes.bytes_left() - if auth { 12 } else { 0 })?),
```

Comment:
I'm pretty sure the only opcode that can be authenticated is `SignedWriteCommand`, so all these `if`s aren't needed.